File: Day22/sortCharactersByFrequency_Donghee.cpp
// Donghee Lee
// 5/22/2020

class Solution {
public:
    // For this algo, Credits to the sample code!! 
    
    // using a map of <int, char>
    string frequencySort(string s) {
      
        int counts[256] = {0};
        
        for (char c : s) {
            counts[c]++;
        }
        
        unordered_map<int, string> mp;
        int maxCount = 0;
        int minCount = INT_MAX;
        
        for (int i = 0; i < 256; i++) {
            if (counts[i] > 0) {
                // appending the char to the string if they have the same counts
                mp[counts[i]] += i;
                 
                // finds the max count
                maxCount = max(maxCount, counts[i]);
                
                // find the min count
                minCount = min(minCount, counts[i]);
            }
        }

        string ans = "";
        
        for (int i = maxCount; i >= minCount; i--) {
            if (mp.find(i) != mp.end()) {
                for (char & c : mp[i]) {
                    for (int j = 0; j < i; j++) {
                        ans += c;
                    }
                }
            }
        }
        
        return ans;
    }
};

// class Solution {
// public:
    
//     // using a priority queue (max heap)
//     string frequencySort(string s) {
//       //  cout << (int)'a' << ", " << (int)'A' << endl;
//         // make a vector for counts for each chars
//         int counts[256] = {0};
        
//         for (char c : s) {
//             counts[(int)c]++;
//         }
        
//         // put elements in the prioity queue to be sorted by their counts
//         priority_queue<pair<int, char>> pq;
//         for (int i = 0; i < 256; i++) {
//             if (counts[i] > 0) {
//                  pq.push(make_pair(counts[i], (char)i));
//             }
//         }

//         string ans = "";
//         while (!pq.empty()) {
//             for (int i = 0; i < pq.top().first; i++) {
//                 ans += pq.top().second;
//             }
//             pq.pop();
//         }
        
//         return ans;
//     }
// };

File: Day22/sortCharactersByFrequency_Donghee.py
import sys

class Solution(object):
    def frequencySort(self, s):
        """
        :type s: str
        :rtype: str
        """
        
        counts = [0 for x in range(256)]

        for c in s:
            counts[ord(c)] += 1

        mp = dict() # mp = { 'a' : 4, 'b' : 2 }

        maxCount = 0
        minCount = sys.maxsize # INT_MAX

        for i in range(256) :
            if counts[i] > 0 :
                if counts[i] not in mp :
                    mp[counts[i]] = ""
                
                mp[counts[i]] += chr(i)
                
                maxCount = max(maxCount, counts[i])
                minCount = min(minCount, counts[i])
        
        ans = ""
        
        # range(start, stop, step)
        for i in range(maxCount, minCount - 1, -1) :
            # if (mp.find(i) != mp.end()) {
            if i in mp : 
                for ch in mp[i] :
                    for j in range(i) :
                        ans += ch

        return ans

            
    # from collections import defaultdict as dd
    # a["a"]; error
    # a['a']; give you a fallback
    # c = Counter('Jeremy')
    # c['e'] -> 2
    # c['a'] -> 0
    # contacts = {}
    # contacts['Donghee'] -> Person()

        


